### PR TITLE
Add system index mappings versions to cluster state

### PR DESCRIPTION
**Pull request by williamrandolph**
Our system index management classes need to access minimum system index mappings version information for the cluster as a whole. There are two reasons for this. First, when creating a system index, we may need to retrieve mappings that are compatible with earlier nodes. Second, and more commonly, we need to know when the minimum system index descriptor version has incremented across all nodes, so that we can update mappings. Currently, we use `Version` for this, but this will move us most of the way towards being able to use mappings versions.

This PR adds the condition to the node join executor that a joining node can't have system index mappings whose version is lower than the minimum version for those mappings on the cluster. In other words, no system index's mapping version can move backwards.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -165,7 +185,8 @@ public TransportVersion read(StreamInput in, String key) throws IOException {
 
     private final Map<String, TransportVersion> transportVersions;
     private final TransportVersion minTransportVersion;
-
+    private final Map<String, VersionsWrapper> versionsWrappers;
+    private final VersionsWrapper minVersions;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java:189]: Could we first generalize this per-node `TransportVersion` map into a `VersionsWrapper` map (containing just the `TransportVersion`) and then add the system index version to `VersionsWrapper` in a separate step? I believe that first step should be possible without changing the transport protocol.

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 2]
```diff
@@ -929,6 +1015,11 @@ public static ClusterState readFrom(StreamInput in, DiscoveryNode localNode) thr
             // infer the versions from discoverynodes for now
             builder.nodes().getNodes().values().forEach(n -> builder.putTransportVersion(n.getId(), inferTransportVersion(n)));
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersion.current())) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/ClusterState.java:1018]: I'd strongly recommend not using `TransportVersion.current()` for this, even in a draft PR - it'd be all too easy to miss one when finalizing things, and CI wouldn't be able to tell you about it until the next transport version bump (by which point the bad version may be released).

**Review by DaveCTurner — COMMENTED** [review]
Makes sense, but I'd like to combine this with `TransportVersion` up-front and ideally do it in a few steps for easier reviews. See inline comments.

File: server/src/main/java/org/elasticsearch/cluster/ClusterState.java (hunk 1 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/cluster/ClusterState.java:189]: I've made a refactoring PR that just wraps TransportVersion with no changes to the transport protocol: https://github.com/elastic/elasticsearch/pull/99114

**Comment by williamrandolph** [comment]
We did this work in a series of more reviewable PRs, leading to https://github.com/elastic/elasticsearch/pull/99307

**Pull request closed without merging**